scripts/gcs_upload.py

```diff
@@ -0,0 +1,578 @@
+import argparse
+import asyncio
+import itertools
+import logging
+import multiprocessing
+import os
+import shutil
+import subprocess
+import tempfile
+import time
+from collections import defaultdict
+from pathlib import PurePath
+
+import google.api_core.exceptions
+import numpy as np
+from dask_jobqueue import SLURMCluster
+from distributed import Client
+from google.cloud.storage import Blob
+from transfer.gcs import create_client, GCSUploader
+from transfer.util.fileutils import collect_filepaths, make_cloud_paths
+
+from cluster.config import load_jobqueue_config
+
+logging.basicConfig(format="%(asctime)s %(message)s", datefmt="%Y-%m-%d %H:%M")
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+
+def _make_boto_options(n_threads):
+    """Create Boto options to pass to gsutil
+    See https://cloud.google.com/storage/docs/boto-gsutil
+    Args:
+        n_threads (int): number of threads to use for parallel uploads
+    Returns:
+        a list of option strings
+    """
+    options = []
+    # Use the same process / thread ratio as the ~/.boto defaults
+    # These defaults are different on Windows (1 process, many threads)
+    processes = max(1, n_threads - 1)
+    options.append(f"Boto:parallel_thread_count={n_threads}")
+    options.append(f"Boto:parallel_process_count={processes}")
+    return options
+
+
+class EnvironmentVariableError(Exception):
+    pass
+
+
+def _set_gsutil_tmpdir():
+    """
+    Sets the temporary directory used by gsutil to the
+    fast node-local directory created for the
+    job this script is running under.
+    """
+    ev = "SLURM_JOB_ID"
+    if ev not in os.environ:
+        raise EnvironmentVariableError(
+            "SLURM_JOB_ID is not set. "
+            "Are you running from a cluster environment?"
+        )
+    job_id = os.environ[ev]
+    if job_id == "":
+        raise EnvironmentVariableError("SLURM_JOB_ID is an empty string")
+    # gsutil writes temporary data to this directory
+    os.environ["TMPDIR"] = f"/scratch/fast/{job_id}/gsutil-tmpdir"
+
+
+async def _rename_blob(bucket, blob, target_path):
+    """
+    Rename the blob to target_path
+    Args:
+        bucket (google.cloud.storage.Bucket) the bucket instance
+        blob (google.cloud.storage.Blob) the blob instance
+        target_path (str) the new name for the blob
+    """
+    try:
+        bucket.rename_blob(blob, target_path)
+    except google.api_core.exceptions.NotFound:
+        # This seems to happen if the target blob already exists??
```

Comment:
can you `logging.warning` this just so it doesn't get completely swallowed?

Reply:
done